// @serverless-stack/resources/dist/util/dotnetBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.builder = void 0;
const chalk_1 = __importDefault(require("chalk"));
const path = __importStar(require("path"));
const fs = __importStar(require("fs-extra"));
const cross_spawn_1 = __importDefault(require("cross-spawn"));
const lambda = __importStar(require("@aws-cdk/aws-lambda"));
const builder_1 = require("./builder");
// Do not re-bundle dependencies for the same srcPath
// Note: if functions from different stacks have the same srcPath, we need to
//       generate the asset for each srcPath, because an asset is tied to a
//       stack. Otherwise, you will get this error:
//
//       Error: Asset is already associated with another stack 'xxxx'.
//              Create a new Code instance for every stack.
const existingBundlesByStackSrcPath = {};
function builder(builderProps) {
    const { stack, srcPath, handler, buildDir } = builderProps;
    const handlerPosixPath = (0, builder_1.getHandlerFullPosixPath)(srcPath, handler);
    console.log(chalk_1.default.grey(`Building Lambda function ${handler} in ${srcPath}`));
    // Check entry path exists
    // Case: invalid srcPath
    if (!fs.existsSync(srcPath)) {
        throw new Error(`Cannot find the srcPath "${srcPath}".`);
    }
    // On Windows, you cannot have ":" in a folder name
    const appPath = process.cwd();
    const handlerHash = (0, builder_1.getHandlerHash)(handlerPosixPath).replace(/::/g, "-");
    const outputPath = path.join(appPath, buildDir, handlerHash);
    // Compile
    let outCode;
    const stackSrcPath = `${stack}/${srcPath}`;
    if (existingBundlesByStackSrcPath[stackSrcPath]) {
        outCode = existingBundlesByStackSrcPath[stackSrcPath];
    }
    else {
        compile();
        outCode = lambda.Code.fromAsset(outputPath);
        existingBundlesByStackSrcPath[stackSrcPath] = outCode;
    }
    return {
        outCode,
        outHandler: handler,
    };
    function compile() {
        const response = cross_spawn_1.default.sync("dotnet", [
            "publish",
            "--output",
            outputPath,
            "--configuration",
            "Release",
            "--framework",
            "netcoreapp3.1",
            "/p:GenerateRuntimeConfigurationFiles=true",
            // warnings are not reported for repeated builds by default and this flag
            // does a clean before build. It takes a little longer to run, but the
            // warnings are consistently printed on each build.
            //"/target:Rebuild",
            "--self-contained",
            "false",
            // do not print "Build Engine version"
            "-nologo",
            // only print errors
            "--verbosity",
            process.env.DEBUG ? "minimal" : "quiet",
        ], {
            stdio: "inherit",
            cwd: path.join(appPath, srcPath),
        });
        if (response.status !== 0) {
            throw new Error(`There was an problem compiling the handler "${handler}" in "${srcPath}".`);
        }
    }
}
exports.builder = builder;
