// @serverless-stack/resources/dist/util/goBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.builder = void 0;
const chalk_1 = __importDefault(require("chalk"));
const path = __importStar(require("path"));
const fs = __importStar(require("fs-extra"));
const cross_spawn_1 = __importDefault(require("cross-spawn"));
const lambda = __importStar(require("@aws-cdk/aws-lambda"));
const builder_1 = require("./builder");
function builder(builderProps) {
    const { srcPath, handler, buildDir } = builderProps;
    const handlerPosixPath = (0, builder_1.getHandlerFullPosixPath)(srcPath, handler);
    console.log(chalk_1.default.grey(`Building Lambda function ${handlerPosixPath}`));
    // Check entry path exists
    if (!fs.existsSync(path.join(srcPath, handler))) {
        throw new Error(`Cannot find a handler file at "${srcPath}/${handler}".`);
    }
    const appPath = process.cwd();
    const handlerHash = (0, builder_1.getHandlerHash)(handlerPosixPath);
    const buildOutputPath = path.join(appPath, buildDir, handlerHash);
    // Compile
    compile(srcPath, handler);
    return {
        outCode: lambda.Code.fromAsset(buildOutputPath),
        outHandler: `handler`,
    };
    function compile(srcPath, handler) {
        const response = cross_spawn_1.default.sync("go", [
            "build",
            "-ldflags",
            "-s -w",
            "-o",
            path.join(buildOutputPath, "handler"),
            // specify absolute path b/c if "handler" can be a folder, and a relative path does not work
            path.join(appPath, srcPath, handler),
        ], {
            stdio: "inherit",
            env: Object.assign(Object.assign({}, process.env), { GOOS: "linux", CGO_ENABLED: "0" }),
            cwd: path.join(appPath, srcPath),
        });
        if (response.status !== 0) {
            throw new Error(`There was an problem compiling the handler at "${srcPath}/${handler}".`);
        }
    }
}
exports.builder = builder;
