// @serverless-stack/resources/dist/util/pythonBuilder.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.builder = void 0;
const chalk_1 = __importDefault(require("chalk"));
const path = __importStar(require("path"));
const fs = __importStar(require("fs-extra"));
const bundling_1 = require("./python/bundling");
const builder_1 = require("./builder");
// Do not re-bundle dependencies for the same srcPath in the same Stack
// Note: if functions from different stacks have the same srcPath, we need to
//       generate the asset for each srcPath, because an asset is tied to a
//       stack. Otherwise, you will get this error:
//
//       Error: Asset is already associated with another stack 'xxxx'.
//              Create a new Code instance for every stack.
const existingBundlesByStackSrcPath = {};
function builder(builderProps) {
    const { stack, bundle, runtime, srcPath, handler } = builderProps;
    const handlerPosixPath = (0, builder_1.getHandlerFullPosixPath)(srcPath, handler);
    console.log(chalk_1.default.grey(`Building Lambda function ${handlerPosixPath}`));
    // Check entry path exists
    const entryPath = path.join(srcPath, (0, builder_1.addExtensionToHandler)(handler, ".py"));
    if (!fs.existsSync(path.join(entryPath))) {
        throw new Error(`Cannot find a handler file at "${entryPath}".`);
    }
    // Bundle dependency with code
    let outCode;
    const stackSrcPath = `${stack}/${srcPath}`;
    if (existingBundlesByStackSrcPath[stackSrcPath]) {
        outCode = existingBundlesByStackSrcPath[stackSrcPath];
    }
    else {
        console.log(chalk_1.default.grey(`Bundling dependencies for ${srcPath} in Docker...`));
        outCode = (0, bundling_1.bundle)({
            installCommands: bundle && bundle.installCommands,
            runtime,
            entry: srcPath,
            outputPathSuffix: ".",
        });
        existingBundlesByStackSrcPath[stackSrcPath] = outCode;
    }
    return {
        outCode,
        outHandler: handler,
    };
}
exports.builder = builder;
