// @serverless-stack/resources/dist/Script/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
const cfnResponse = __importStar(require("./cfn-response"));
const outbound_1 = require("./outbound");
const util_1 = require("./util");
function handler(cfnRequest) {
    return __awaiter(this, void 0, void 0, function* () {
        util_1.log("onEventHandler", cfnRequest);
        // Invoke user function on Create and on Update
        const fnCreate = cfnRequest.ResourceProperties.UserCreateFunction;
        const fnUpdate = cfnRequest.ResourceProperties.UserUpdateFunction;
        const fnDelete = cfnRequest.ResourceProperties.UserDeleteFunction;
        const fnParams = JSON.parse(cfnRequest.ResourceProperties.UserParams);
        if (cfnRequest.RequestType === "Create" && fnCreate) {
            yield invokeUserFunction(fnCreate, { params: fnParams });
        }
        else if (cfnRequest.RequestType === "Update" && fnUpdate) {
            yield invokeUserFunction(fnUpdate, { params: fnParams });
        }
        else if (cfnRequest.RequestType === "Delete" && fnDelete) {
            yield invokeUserFunction(fnDelete, { params: fnParams });
        }
        // Build response
        return cfnResponse.submitResponse("SUCCESS", Object.assign(Object.assign({}, cfnRequest), { PhysicalResourceId: defaultPhysicalResourceId(cfnRequest) }));
    });
}
function invokeUserFunction(functionName, payload) {
    return __awaiter(this, void 0, void 0, function* () {
        util_1.log(`executing user function ${functionName} with payload`, payload);
        const resp = yield outbound_1.invokeFunction({
            FunctionName: functionName,
            Payload: JSON.stringify(payload),
        });
        util_1.log("user function response:", resp, typeof resp);
        if (resp.FunctionError) {
            util_1.log("user function threw an error:", resp.FunctionError);
            const jsonPayload = parseJsonPayload(resp.Payload);
            // Note: custom resources have a response limit of 4k. Limit the
            //       error message to 1000 characters.
            const errorMessage = (jsonPayload.errorMessage || "error").substring(0, 1000);
            // append a reference to the log group.
            const message = [
                errorMessage,
                "",
                `Logs: /aws/lambda/${functionName}`,
                "",
            ].join("\n");
            const e = new Error(message);
            // the output that goes to CFN is what's in `stack`, not the error message.
            // if we have a remote trace, construct a nice message with log group information
            if (jsonPayload.trace) {
                // skip first trace line because it's the message
                e.stack = [message, ...jsonPayload.trace.slice(1)].join("\n");
            }
            throw e;
        }
    });
}
function parseJsonPayload(payload) {
    if (!payload) {
        return {};
    }
    const text = payload.toString();
    try {
        return JSON.parse(text);
    }
    catch (e) {
        throw new Error(`return values from user-handlers must be JSON objects. got: "${text}"`);
    }
}
function defaultPhysicalResourceId(req) {
    switch (req.RequestType) {
        case "Create":
            return req.RequestId;
        case "Update":
        case "Delete":
            return req.PhysicalResourceId;
        default:
            throw new Error(`Invalid "RequestType" in request "${JSON.stringify(req)}"`);
    }
}
module.exports = {
    handler: cfnResponse.safeHandler(handler),
};
