// @serverless-stack/resources/dist/Construct.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCDKConstruct = exports.isCDKConstructOf = exports.isSSTConstruct = exports.isStackConstruct = exports.isConstruct = exports.getFunctionRef = void 0;
const aws_lambda_1 = require("aws-cdk-lib/aws-lambda");
const Stack_1 = require("./Stack");
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
function getFunctionRef(fn) {
    if (!fn)
        return undefined;
    if (!(fn instanceof aws_lambda_1.Function))
        return undefined;
    return {
        node: fn.node.addr,
        stack: Stack_1.Stack.of(fn).node.id,
    };
}
exports.getFunctionRef = getFunctionRef;
function isConstruct(construct) {
    return isSSTConstruct(construct) || isCDKConstruct(construct);
}
exports.isConstruct = isConstruct;
function isStackConstruct(construct) {
    return isCDKConstructOf(construct, "aws-cdk-lib.Stack");
}
exports.isStackConstruct = isStackConstruct;
function isSSTConstruct(construct) {
    return "getConstructMetadata" in construct;
}
exports.isSSTConstruct = isSSTConstruct;
function isCDKConstructOf(construct, moduleName) {
    var _a, _b;
    // We need to check if construct is an CDK construct. To do that:
    // - we cannot use the `construct instanceof` check because ie. the PolicyStatement
    //   instance in the user's app might come from a different npm package version
    // - we cannot use the `construct.constructor.name` check because the constructor
    //   name can be prefixed with a number ie. PolicyStatement2
    //
    // Therefore we are going to get the constructor's fqn. The constructor for a CDK
    // construct looks like:
    //    [class Bucket2 extends BucketBase] {
    //      [Symbol(jsii.rtti)]: { fqn: '@aws-cdk/aws-s3.Bucket', version: '1.91.0' }
    //    }
    // We will check against `fqn`.
    const fqn = (_b = (_a = construct === null || construct === void 0 ? void 0 : construct.constructor) === null || _a === void 0 ? void 0 : _a[JSII_RTTI_SYMBOL_1]) === null || _b === void 0 ? void 0 : _b.fqn;
    return typeof fqn === "string" && fqn === moduleName;
}
exports.isCDKConstructOf = isCDKConstructOf;
function isCDKConstruct(construct) {
    var _a, _b;
    const fqn = (_b = (_a = construct === null || construct === void 0 ? void 0 : construct.constructor) === null || _a === void 0 ? void 0 : _a[JSII_RTTI_SYMBOL_1]) === null || _b === void 0 ? void 0 : _b.fqn;
    return (typeof fqn === "string" &&
        (fqn.startsWith("@aws-cdk/") || fqn.startsWith("aws-cdk-lib")));
}
exports.isCDKConstruct = isCDKConstruct;
