// @serverless-stack/resources/dist/ViteStaticSite.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ViteStaticSite = void 0;
const path = __importStar(require("path"));
const fs = __importStar(require("fs-extra"));
const StaticSite_1 = require("./StaticSite");
/////////////////////
// Construct
/////////////////////
class ViteStaticSite extends StaticSite_1.StaticSite {
    constructor(scope, id, props) {
        const { path: sitePath, environment, typesPath } = props || {};
        // generate buildCommand
        let defaultBuildCommand = "npm run build";
        if (fs.existsSync(path.join(sitePath, "yarn.lock"))) {
            defaultBuildCommand = "yarn build";
        }
        // create types file
        const defaultTypesPath = "src/sst-env.d.ts";
        if (typesPath === undefined && isTs(sitePath)) {
            const filePath = path.resolve(path.join(sitePath, defaultTypesPath));
            generateTypesFile(filePath, environment);
        }
        else if (typeof typesPath === "string") {
            const filePath = path.resolve(path.join(sitePath, typesPath));
            generateTypesFile(filePath, environment);
        }
        super(scope, id, Object.assign({ indexPage: "index.html", errorPage: StaticSite_1.StaticSiteErrorOptions.REDIRECT_TO_INDEX_PAGE, buildCommand: defaultBuildCommand, buildOutput: "dist", fileOptions: [
                {
                    exclude: "*",
                    include: "*.html",
                    cacheControl: "max-age=0,no-cache,no-store,must-revalidate",
                },
                {
                    exclude: "*",
                    include: ["*.js", "*.css"],
                    cacheControl: "max-age=31536000,public,immutable",
                },
            ] }, props));
    }
}
exports.ViteStaticSite = ViteStaticSite;
function isTs(sitePath) {
    return fs.existsSync(path.join(sitePath, "tsconfig.json"));
}
function generateTypesFile(typesFullPath, environment) {
    const content = `/// <reference types="vite/client" />

interface ImportMetaEnv {
${Object.keys(environment || {}).map(key => `  readonly ${key}: string`).join("\n")}
}

interface ImportMeta {
readonly env: ImportMetaEnv
}`;
    fs.ensureDirSync(path.dirname(typesFullPath));
    fs.writeFileSync(typesFullPath, content);
}
